File: oc/UIPickerViewDemo/UIPickerViewDemo/FoodViewController.m
//
//  FoodViewController.m
//  UIPickerViewDemo
//
//  Created by dfang on 2018-1-6.
//  Copyright © 2018年 east. All rights reserved.
//

#import "FoodViewController.h"

@interface FoodViewController ()

@property (nonatomic, strong) UIPickerView *pickerView;
@property (nonatomic, strong) NSArray *foods;
@property (nonatomic, strong) UILabel *fruitLabel;
@property (nonatomic, strong) UILabel *foodLabel;
@property (nonatomic, strong) UILabel *drinkLabel;
@property (nonatomic, strong) UIButton *randomButton;
@property (nonatomic, strong) UILabel *titleLabel;

@end

@implementation FoodViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.pickerView.delegate = self;
    self.pickerView.dataSource = self;
    [self.view addSubview:self.pickerView];
    
    [self.view addSubview:self.titleLabel];
    [self.view addSubview:self.randomButton];
    [self.view addSubview:self.fruitLabel];
    [self.view addSubview:self.foodLabel];
    [self.view addSubview:self.drinkLabel];
    
    for (int i = 0; i < self.foods.count; i++) {
        [self pickerView:self.pickerView didSelectRow:0 inComponent:i];
    }
}

- (UIPickerView *)pickerView {
    if (_pickerView == nil) {
        _pickerView = [[UIPickerView alloc] initWithFrame:CGRectMake(0, 100, self.view.frame.size.width, 200)];
    }
    
    return _pickerView;
}

- (NSArray *)foods {
    if (_foods == nil) {
        NSString *foodFile = [[NSBundle mainBundle] pathForResource:@"foods" ofType:@"plist"];
        _foods = [NSArray arrayWithContentsOfFile:foodFile];
    }
    return _foods;
}

- (UILabel *)fruitLabel {
    if (_fruitLabel == nil) {
        _fruitLabel = [[UILabel alloc] initWithFrame:CGRectMake(50, 310, self.view.frame.size.width, 50)];
        _fruitLabel.text = @"水果：";
    }
    return _fruitLabel;
}

- (UILabel *)foodLabel {
    if (_foodLabel == nil) {
        _foodLabel = [[UILabel alloc] initWithFrame:CGRectMake(50, 360, self.view.frame.size.width, 50)];
        _foodLabel.text = @"主食：";
    }
    return _foodLabel;
}

- (UILabel *)drinkLabel {
    if (_drinkLabel == nil) {
        _drinkLabel = [[UILabel alloc] initWithFrame:CGRectMake(50, 410, self.view.frame.size.width, 50)];
        _drinkLabel.text = @"饮料：";
    }
    return _drinkLabel;
}

- (UIButton *)randomButton {
    if (_randomButton == nil) {
        _randomButton = [[UIButton alloc] initWithFrame:CGRectMake(10, 50, 100, 50)];
        [_randomButton setTitle:@"随机" forState:UIControlStateNormal];
        [_randomButton setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        [_randomButton addTarget:self action:@selector(randomBtnClicked) forControlEvents:UIControlEventTouchUpInside];
    }
    return _randomButton;
}

- (UILabel *)titleLabel {
    if (_titleLabel == nil) {
        _titleLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 50, self.view.frame.size.width, 50)];
        _titleLabel.text = @"点菜系统";
        _titleLabel.textAlignment = NSTextAlignmentCenter;
    }
    return _titleLabel;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSInteger)numberOfComponentsInPickerView:(nonnull UIPickerView *)pickerView {
    return self.foods.count;
}

- (NSInteger)pickerView:(nonnull UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    NSArray *items = self.foods[component];
    return items.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    NSArray *items = self.foods[component];
    return  items[row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    NSArray *items = self.foods[component];
    if (component == 0) {
        self.fruitLabel.text = [NSString stringWithFormat:@"水果：%@",  items[row]];
    } else if (component == 1) {
        self.foodLabel.text = [NSString stringWithFormat:@"主食：%@",  items[row]];
    } else if (component == 2) {
        self.drinkLabel.text = [NSString stringWithFormat:@"饮料：%@",  items[row]];
    }
}

- (void)randomBtnClicked {
    for (int i = 0; i < self.foods.count; i++) {
        NSArray *items = self.foods[i];
        
        NSInteger oldSelected = [self.pickerView selectedRowInComponent:i];
        NSInteger randomSelected = arc4random_uniform(items.count);
        while (randomSelected == oldSelected) {
           randomSelected = arc4random_uniform(items.count);
        }
        
        [self.pickerView selectRow:randomSelected inComponent:i animated:true];
        [self pickerView:self.pickerView didSelectRow:randomSelected inComponent:i];
    }
}

@end

File: swift/UIPickerViewDemo/UIPickerViewDemo/FoodViewController.swift
//
//  FoodViewController.swift
//  UIPickerViewDemo
//
//  Created by dfang on 2018-1-5.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class FoodViewController: UIViewController, UIPickerViewDataSource, UIPickerViewDelegate {
    
    let pickerView:UIPickerView = UIPickerView.init()
    
    let headerView = UIView.init()
    let foods = NSArray.init(contentsOfFile: Bundle.main.path(forResource: "foods", ofType: "plist")!)!
    
    let fruitLabel = UILabel.init()
    let foodLabel = UILabel.init()
    let drinkLabel = UILabel.init()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        let screenWidth = UIScreen.main.bounds.width
        headerView.frame = CGRect.init(x: 0, y: 0, width: screenWidth, height: 64)
        let titleLabel = UILabel.init(frame: headerView.frame)
        titleLabel.text = "点菜系统"
        titleLabel.font = UIFont.boldSystemFont(ofSize: 20)
        titleLabel.textAlignment = NSTextAlignment.center
        headerView.addSubview(titleLabel)
        
        let randomBtn = UIButton.init(frame: CGRect.init(x: 10, y: 20, width: 80, height: 50))
        randomBtn.setTitle("随机", for: UIControlState.normal)
        randomBtn.setTitleColor(UIColor.blue, for: UIControlState.normal)
        randomBtn.addTarget(self, action: #selector(randomBtnClicked), for: UIControlEvents.touchUpInside)
        headerView.addSubview(randomBtn)
    
        self.view.addSubview(headerView)
        
        pickerView.frame = CGRect.init(x: 0, y: 80, width: screenWidth, height: 200)
        pickerView.dataSource = self
        pickerView.delegate = self
        self.view.addSubview(pickerView)
        
        fruitLabel.frame = CGRect.init(x: 20, y: 300, width: 250, height: 50)
        self.view.addSubview(fruitLabel)
        foodLabel.frame = CGRect.init(x: 20, y: 350, width: 250, height: 50)
        self.view.addSubview(foodLabel)
        drinkLabel.frame = CGRect.init(x: 20, y: 400, width: 250, height: 50)
        self.view.addSubview(drinkLabel)
        
        for i in 0...(foods.count-1) {
            self.pickerView(pickerView, didSelectRow: 0, inComponent: i)
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return foods.count
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return (foods[component] as! NSArray).count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        let items = foods[component] as! NSArray
        return items[row] as? String
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        let items = foods[component] as! NSArray
        let item = items[row] as! String
        switch component {
        case 0:
            fruitLabel.text = "水果：" + item
            break
        case 1:
            foodLabel.text = "主食：" + item
            break
        case 2:
            drinkLabel.text = "饮料：" + item
        default:
            break
        }
    }
    
    @objc func randomBtnClicked() {
        for i in 0...(foods.count-1) {
            let items = foods[i] as! NSArray
            var randomRow = Int(arc4random_uniform(UInt32(items.count)))
            let oldRow = pickerView.selectedRow(inComponent: i)
            while (randomRow == oldRow) {
                randomRow = Int(arc4random_uniform(UInt32(items.count)))
            }
            
            // 更改pickerview中显示数据
            pickerView.selectRow(randomRow, inComponent: i, animated: true)
            // 更改label显示数据
            self.pickerView(pickerView, didSelectRow: randomRow, inComponent: i)
        }
    }
}

